src/distributed-callback-queue.js

```diff
@@ -195,10 +195,23 @@ class DistributedCallbackQueue {
     const timeout = props.pop();
     assert(typeof worker === 'function', 'ensure that you pass a function as a worker');
 
+    // allows us to reject-and-halt (eg. on timeout) even if the #push'ed lock has not yet been acquired
+    let jobAbortReject;
+    let jobAbortPromise = new Promise((resolve, reject) => {
+      jobAbortReject = reject;
+    });
+
     let onJobCompleted;
-    const promise = new Promise((resolve, reject) => {
+    const jobCompletedPromise = new Promise((resolve, reject) => {
       onJobCompleted = (err, ...args) => {
         if (err) {
+          if (jobAbortPromise) {
+            // ensure that jobAbortPromise rejects *first* so that we can return jobCompletedPromise *before* it rejects
+            jobAbortReject(err);
+            setImmediate(reject, err);
```

Comment:
ensure that the `jobCompletedPromise` gets returned *before* it rejects, or the 'unhandledRejection' will still occur

rejecting `jobAbortPromise` immediately will resolve the `Promise.race` and allow `jobCompletedPromise` be returned synchronously to the caller for proper handling before it self-rejects "in just a moment"